core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java

```diff
@@ -417,8 +450,8 @@ public IteratorResponse startIterator(Object requestId, Address origin, IntSet s
       }
       CacheSet<Original> originalSet = toOriginalSet(includeLoader, entryStream);
 
-      // Have to make a copy of segments as we will remove from it below
-      SegmentListener listener = new SegmentListener(IntSets.mutableCopyFrom(segments), i -> true);
+      // Unfortunately we have to ALWAYS use a listener - due to storing state between requests in the listener
```

Comment:
IMO this isn't very helpful for readers, because it answers one question that I didn't think of asking (why do we use a listener here) and asks more questions for which I now need to interrupt my flow and find the answers somewhere else:

1. Why wouldn't we use a listener?
2. What kind of data do we store in the listener?
3. Can multiple iteration requests share a listener?

Reply:
Sadly this another one of those shut down things. I can try to explain this one a bit better in comments. It is because of the lazily marshalled approach it takes when sending a response.